lib/charms/traefik_route_k8s/v0/traefik_route.py

```diff
@@ -140,25 +144,42 @@ def __init__(self, charm: CharmBase, relation_name: str = 'traefik-route'):
                 (defaults to "traefik-route").
         """
         super().__init__(charm, relation_name)
+        self._stored.set_default(external_host=None)
+
         self.charm = charm
-        self.framework.observe(self.charm.on[relation_name].relation_changed,
-                               self._on_relation_changed)
+        self._relation_name = relation_name
+
+        if self._stored.external_host != charm.external_host:
+            self._stored.external_host = charm.external_host
+            self._update_requirers_with_external_host()
```

Comment:
In general, I think it is best to avoid StoredState wherever possible. I see no reason here why we can't just always call `_update_requirers_with_external_host`

Reply:
`StoredState` is generally avoided for complex datastructures for a variety of reasons, but this is not, and it [matches what traefik already does](https://github.com/PietroPasotti/traefik-route-k8s-operator/blob/main/lib/charms/traefik_k8s/v0/ingress.py#L357). Sometimes, writing patches means using the conventions they already use.

The reason not to blindly update relation data every time something happens is that, generally, we are asking Juju to do additional work in verifying that nothing has actually changed (to not propagate that event), and in the worst case, if that contract is ever changed or it was a never contract at all, every single time anything happens, we will tell the Provider that there's a new external host, and it will probably reconfigure/restart itself, causing a workflow interruption.

Reply:
Consumers will only be notified of a change if one actually occurs.

In terms of juju needing to "do work" to check if we have changed anything, that work is negligible and possibly not even more than fetching a stored state anyway.